HTML search: serialize search index as JSON

jayaddison opened the pull request:
### Feature or Bugfix
- Bugfix

### Purpose
- Prevent queries for the JavaScript class prototype reference propery name from causing errors in the HTML search JavaScript code provided in the `basic` theme.

### Detail
- Ensure that the user input query term is found as a named property of the relevant index objects before accessing them, using [`Object.hasOwnProperty`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty).
- ~~Represent the HTML search indices as [JavaScript `Map`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) instances instead of `Object` literals.~~
- Serialize the search index content as JSON, so that terms named `__proto__` can be serialized and loaded.

### Relates
- Resolves #13096.
- ~~May resolve # 13098 (conditional on inclusion of `JSON.parse` reviver / `Object.freeze` logic)~~

Edit: update description to reflect JavaScript Map-based approach.
Edit: update description to reflect JSON-serialization approach.

jayaddison committed:
search: check that query terms exist as properties in term indices before accessing them

jayaddison committed:
Tests: fixtures: add document title for ECMAScript testroot

jayaddison committed:
Tests: add JavaScript test coverage for prototype-property-name query

jayaddison committed:
Add CHANGES.rst entry

[skip ci]

wlach commented on sphinx/themes/basic/static/searchtools.js:
```diff
@@ -514,8 +514,8 @@ const Search = {
       const arr = [
-        { files: terms[word], score: Scorer.term },
-        { files: titleTerms[word], score: Scorer.title },
+        { files: terms.hasOwnProperty(word) ? terms[word] : [], score: Scorer.term },
```
I'm actually a little confused by how this might work, as it doesn't appear that JavaScript supports setting the `__proto__` property in an object. Am I missing something? Should we be using a map to store these terms instead?

jayaddison replied:
When parsing JSON, an object key named `"__proto__"` is accepted, and converted to the JavaScript object key `["__proto__"]` -- distinct from the `__proto__` property (a reference to a class prototype).

During read-access, introducing `hasOwnProperty` allows us to filter out the latter, so that we don't mistakenly retrieve the prototype class reference from `terms`.

I'm not certain whether the below example will help to clarify, but I find experimentation helpful to understand some of the behaviours:

```javascript
> x = JSON.parse('{}')
{}
> y = JSON.parse('{"__proto__": "testing"}')
{ ['__proto__']: 'testing' }
> '__proto__' in x
true
> x.hasOwnProperty('__proto__')
false
> '__proto__' in y
true
> y.hasOwnProperty('__proto__')
true
```

jayaddison replied:
I forgot to respond to your `Map` question: that is certainly possible, although I feel that doing so might add some runtime overhead (since we'd need to load the native JS object contents from the index into respective `Map` objects).

As I understand it, `Map` objects are also more difficult to make immutable in JavaScript than `Object` and `Array` instances.  Freezing a `Map` object doesn't prevent its private contents from being updated:

```javascript
> x = {'foo': 'bar'}
{ foo: 'bar' }
> Object.freeze(x)
{ foo: 'bar' }
> x['foo'] = 'baz'
'baz'
> x['foo']
'bar'
> y = new Map([['foo', 'bar']])
Map(1) { 'foo' => 'bar' }
> Object.freeze(y)
Map(1) { 'foo' => 'bar' }
> y.set('foo', 'baz')
Map(1) { 'foo' => 'baz' }
> y.get('foo')
'baz'
```

So another factor is that, to the extent that I'd like to see the search index contents become immutable at load-time in future (#13098), I'd prefer not to migrate to the `Map` type unless there is a compelling reason to (maybe avoiding prototype pollution is that reason -- but I don't think we're exposed to that currently).

wlach replied:
Ok got it, this makes sense to me re: the property. TIL!

I'm a little bit less sure about whether the index should be immutable, but happy to defer that discussion to another day.

wlach replied:
Actually, one extra piece of optional feedback: since this is kind of counter-intuitive, you might want to add a comment on this line explaining what you're doing.

jayaddison replied:
Yep, OK - I'll try to add some phrasing to explain this succinctly.  Thanks for the code review!

wlach commented on tests/js/roots/ecmascript/index.rst:
```diff
@@ -0,0 +1,8 @@
+
+This is a sample JavaScript (aka ``ECMAScript``) project used to generate a search engine index fixture.
+
+.. js:attribute:: Object.__proto__
```
Should we test this on a file *without* `__proto__` in it to make sure we don't match?

jayaddison replied:
That probably does make sense too, yep - thanks :+1:

jayaddison replied:
(in fact -- no probably about it -- adding a test like that definitely does make sense.  ensuring that no results are mistakenly reported when the term `__proto__` does not exist in the term index is the core of the bug)

wlach reviewed (commented):
I can see how this would be a problem but I'm a little unclear on how the solution actually works? See comments

jayaddison committed:
Tests: add negative-test case

The JS prototype property should _not_ be found in unrelated documents

jayaddison commented on sphinx/themes/basic/static/searchtools.js:
```diff
@@ -514,8 +514,8 @@ const Search = {
-        { files: terms[word], score: Scorer.term },
-        { files: titleTerms[word], score: Scorer.title },
+        { files: terms.hasOwnProperty(word) ? terms[word] : [], score: Scorer.term },
+        { files: titleTerms.hasOwnProperty(word) ? titleTerms[word] : [], score: Scorer.title },
```
Preparing that comment has made me realize: we perform an `===` equality check against the `undefined` value further down the function.  We should probably not alter that behaviour in this PR.

jayaddison committed:
search: add explanatory comment for document title/text term lookup

jayaddison committed:
search: retain `undefined` result instead of `[]` for non-matching terms

jayaddison committed:
search: add explanatory comment regarding `hasOwnProperty` usage

AA-Turner committed:
Merge branch 'master' into issue-13096/html-search-handle-proto-query

jayaddison committed:
HTML search: encapsulate the search index content within a JSON string

jayaddison committed:
HTML search: use the `JSON.parse` `reviver` argument to freeze search index contents

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

jayaddison committed:
Merge branch 'master' into issue-13096/html-search-handle-proto-query

jayaddison committed:
Tests: fixup: accommodate adjusted `setIndex` prefix/suffix lengths

jayaddison committed:
HTML search: replace `repr` with simple embedded single-quotes

The `JSON.dumps` call escapes single-quotes within strings; all JSON strings are enclosed by double-quotes

AA-Turner commented on sphinx/themes/basic/static/searchtools.js:
```diff
@@ -219,7 +219,8 @@ const Search = {
   setIndex: (index) => {
-    Search._index = index;
+    const reviver = (k, v) => (typeof v === "object" && v !== null) ? Object.freeze(v) : v;
+    Search._index = JSON.parse(index, reviver);
```
I think @wlach is right that immutability is orthoganal here. I think 'simply' using `JSON.parse` rather than an object literal would solve the `__proto__` case. Would we then need to keep the `hasOwnProperty` checks?

An alternative would be to use a reviver function that constructs `Map` objects/serialise the index with `Map` literals, and use `.get()`, but that may also be expensive.

```suggestion
    Search._index = JSON.parse(index);
```

jayaddison replied:
> I think @wlach is right that immutability is orthoganal here. I think 'simply' using `JSON.parse` rather than an object literal would solve the `__proto__` case.

Ok, yep - I can revert 443fd213c75d5ba7eb9b04869ccad8ca655f2393 to remove the use of `Object.freeze` / the reviver function.
 
>  Would we then need to keep the `hasOwnProperty` checks?

We should keep those, I think.  Without them, queries for `__proto__` may continue to inadvertently retrieve object prototypes instead of getting an `undefined` result.

> An alternative would be to use a reviver function that constructs `Map` objects/serialise the index with `Map` literals, and use `.get()`, but that may also be expensive.

What do you think about constructing those `Map` objects directly in the `searchindex.js` file format, instead of using `JSON.parse` and a reviver?

wlach replied:
> What do you think about constructing those Map objects directly in the searchindex.js file format, instead of using JSON.parse and a reviver?

IMO this is definitely worth a shot, I suspect it'll be quite fast.

jayaddison replied:
I'm giving it a try - it doesn't seem too tricky; the Python serialization of the search index becomes a little more complex.  One challenge that I have yet to work out is how the Python unit tests should parse/load the resulting index (ref: https://github.com/sphinx-doc/sphinx/blob/15b46dc81e56ac1fd352da808f7aadd3e825a101/tests/test_search.py#L70).

jayaddison committed:
Revert "HTML search: use the `JSON.parse` `reviver` argument to freeze search index contents"

This reverts commit 443fd213c75d5ba7eb9b04869ccad8ca655f2393.

jayaddison committed:
WIP: HTML search: use JS `Map`s in serialized index

[skip ci]

jayaddison committed:
HTML search: fixup for object search

Relates-to commit d4bba1ea1f6cd3730fc8891d9bcc65fb7f309f31.

[skip ci]

jayaddison committed:
HTML search: ensure array-format index entries are sorted

jayaddison committed:
HTML search: refactor-out redundant code

jayaddison committed:
HTML search: rectify variable names

jayaddison committed:
HTML search: compress index representation

jayaddison committed:
HTML search: enclose index names in double-quotes

jayaddison committed:
Tests: HTML search: use `ast.literal_eval` to parse index contents

jayaddison committed:
Tests: HTML search: update test expectation

jayaddison committed:
HTML search: cleanup: remove redundant `assert` statement

jayaddison committed:
HTML search: typing/lint fixups for `ruff` and `mypy`

jayaddison commented on tests/test_search.py:
```diff
@@ -67,7 +68,16 @@ def load_searchindex(path: Path) -> Any:
+        .replace('new Map', '')
+        .replace('null', 'None')
+        .replace('false', 'False')
+    )
```
This is fragile and inelegant; it's a compromise to avoid introducing the complexity of a Python-implemented JavaScript parser dependency given the use of `new Map(...)` within `searchindex.js` files.

wlach replied:
Another alternative would be to dump the index as JSON, then reload it into the actual format we want. Would almost certainly be a little slower, but probably not disastrously so.

jayaddison replied:
Could you explain that in a little more detail?  I think I understand what you mean, but I'd like to make sure.

wlach replied:
Yeah I mean actually just writing out a JSON file and having the frontend code load it seperately via a fetch call. I'm not actually sure it would be slower, now that I think about it-- this post suggests this will be faster on the interpreter side: https://v8.dev/blog/cost-of-javascript-2019#json

jayaddison replied:
Ok, thanks! (I wasn't entirely sure whether you were suggesting JSON format output only for the unit test code - using it in both seems more robust).

Let's give `searchindex.json` a try, in that case.  It'll make a few things simpler, although the `searchtools.js` code may lose some of the readability improvements (we'll want to begin/reintroduce using `obj.hasOwnProperty` again in a few places).

We may also want to be careful about where we place the `fetch` call so that the browser doesn't block for unnecessarily long before invoking that (perhaps especially relevant for the case of downstream themes that include their own scripting that may also run on each page load).

wlach replied:
> Let's give searchindex.json a try, in that case. It'll make a few things simpler, although the searchtools.js code may lose some of the readability improvements (we'll want to begin/reintroduce using obj.hasOwnProperty again in a few places).

Personally I would read the search index as JSON, then convert it into a Map (i.e. building on the work you already did).

> We may also want to be careful about where we place the fetch call so that the browser doesn't block for unnecessarily long before invoking that (perhaps especially relevant for the case of downstream themes that include their own scripting that may also run on each page load).

Yeah that part might be fiddly. Some types of interactivity might not be fully supported until this operation is complete, so this might complicate things. If it gets too out of control, maybe best to go back to the previous approach.

jayaddison replied:
> > Let's give searchindex.json a try, in that case. It'll make a few things simpler, although the searchtools.js code may lose some of the readability improvements (we'll want to begin/reintroduce using obj.hasOwnProperty again in a few places).
> 
> Personally I would read the search index as JSON, then convert it into a Map (i.e. building on the work you already did).

I'd prefer to avoid multiple read passes over the search index data on the client-side, if possible.  Similarly that makes me reluctant to use the `reviver` function in `JSON.parse`, and I considered it a downside of the freeze-and-deproto code I had implemented in #13098.

Aside from code clarity, do you anticipate many benefits from the transformation to `Map` datastructures?

wlach replied:
> Aside from code clarity, do you anticipate many benefits from the transformation to Map datastructures?

Not really, my guess is that getting entries out of a map will be a little faster (since it's a datastructure purely oriented around storing and retrieving keys/values) but probably not noticeably so. I think it's at least worth exploring though. The conversion should just be a few extra lines of code after you've loaded the JSON structure.

jayaddison replied:
Mm, I agree it'd only be a few lines of code, but I don't think that it'd be worth the resulting compute resource expenditure to do that.

jayaddison replied:
Generally I feel inclined to disentangle the two aspects of this pull request: adding support for `__proto__` as an indexed/queryable term, and preventing `__proto__` queries from accessing JavaScript object prototypes.

The latter seems more important (it's a bug) - the former is a feature request that hasn't really been requested by anyone other than myself during development.

jayaddison replied:
I've extracted the bugfix part of this changeset into #13153; I don't have any plans currently to add feature support for term/title-term matching of the query term `__proto__` itself.

jayaddison committed:
Edit CHANGES.rst entry

[skip ci]

wlach commented on sphinx/search/__init__.py:
```diff
@@ -168,8 +168,17 @@ class _JavaScriptIndex:
+            for name, entries in sorted(data.items())
+        }
+        data_js = '{' + ','.join(f'{k}:{v}' for k, v in js_indices.items()) + '}'
+        return self.PREFIX + data_js + self.SUFFIX
```
Could we use a [jinja](https://jinja.palletsprojects.com/en/stable/) template for the search index? That might make this look a little better. Seems to be one of Sphinx's dependencies: https://github.com/sphinx-doc/sphinx/blob/80ed1b4cd1a5290106f2bff528b27f0bf16fb32a/pyproject.toml#L65

Definitely worth adding a comment or two explaining what's going on.

jayaddison replied:
I'm mulling over the suggestion; making the logic and output format of the method clearer would both be helpful, but I'd probably prefer not to include an additional technology in the mix (even though as you mention, it's already available as an application dependency).

wlach replied:
I think we should use it if it's available and it makes things clearer. It's usage elsewhere in Sphinx is pretty extensive: https://github.com/search?q=repo%3Asphinx-doc%2Fsphinx%20jinja&type=code

wlach replied:
(although dumping the index as JSON is perhaps a better alternative)

AA-Turner replied:
@wlach I'm fairly sure Jinja was originally written for use in Sphinx / Flask -- they're all originally Pocco projects!

wlach commented on sphinx/themes/basic/static/searchtools.js:
```diff
@@ -333,7 +333,7 @@ const Search = {
 
     const queryLower = query.toLowerCase().trim();
-    for (const [title, foundTitles] of Object.entries(allTitles)) {
+    for (const [title, foundTitles] of allTitles) {
```
This (the whole file) actually looks a lot better to my eye.

jayaddison replied:
Yep, I'd agree with that.  Ideally I'd prefer if we could use null-prototype objects so that the presence-tests can become `x in obj` instead of `obj.hasOwnProperty(x)` -- but perhaps that'll be possible again in future.

wlach reviewed (commented):
> There is a small amount of index size bloat -- not so much from the new Map( and ) text overhead, but more from the additional Array literals (e.g. square brackets) used to format their entries instead of object literals. This is approximately 5% overhead on the test dataset; 541857 bytes compared to 516691 baseline.

My guess is that most of this difference (which is relatively small) would disappear with compression.

> There seem to be multiple factors to balance: Python complexity, JavaScript complexity, code safety (both datastructure immutability, and also prototype access), runtime performance and index filesize.

Yep, these things are always about tradeoffs.

> I'm not sure how much to explore and how to present my thoughts/findings. Generally my feeling is that the Map approach here (again, up to https://github.com/sphinx-doc/sphinx/commit/fae81a0a92e81618c8fc70b2db1f3f93bc0dccaa) seems good on most criteria except for Python complexity (particularly the unit tests, but also the application serialization code is more complex than it was) and immutability (that it does not address, in my opinion). The tc39 Records-and-Tuples proposal, I think, would be near-optimal on all factors -- but we could be years away from that reaching widespread adoption.

I think it's a pretty good solution. I'll defer approving in case you want to experiment / document more, but I think this fixes the bug while still being reasonable to maintain.

jayaddison committed:
HTML search: use JSON format for the search index

jayaddison committed:
Merge branch 'master' into issue-13096/html-search-handle-proto-query

Conflicts:
	CHANGES.rst

jayaddison committed:
Tests: cleanup: remove unused `ast` import

jayaddison committed:
HTML search: refactor loader to use Fetch API

Ref: https://fetch.spec.whatwg.org/#fetch-api
